signer/tests/integration/containers.rs

```diff
@@ -0,0 +1,305 @@
+    async fn get_bitcoin_url(&self) -> Result<Url, Error> {
+        let host = self.containers.get_service_host(SERVICE_BITCOIN).await?;
+        let port = self
+            .containers
+            .get_service_port(SERVICE_BITCOIN, SERVICE_BITCOIN_RPC_PORT)
+            .await?;
+        format!(
+            "http://{}:{}@{}:{}",
+            BITCOIN_CORE_RPC_USERNAME, BITCOIN_CORE_RPC_PASSWORD, host, port
+        )
+        .parse()
+        .map_err(Error::InvalidUrl)
+    }
+
+    /// Get the database address
+    async fn get_database_url(&self) -> Result<String, Error> {
+        let host = self.containers.get_service_host(SERVICE_DATABASE).await?;
+        let port = self
+            .containers
+            .get_service_port(SERVICE_DATABASE, SERVICE_DATABASE_PORT)
+            .await?;
+        Ok(format!(
+            "postgres://{}:{}@{}:{}",
+            SERVICE_DATABASE_USER, SERVICE_DATABASE_PASSWORD, host, port
+        ))
+    }
+
+    /// Get the Bitcoin container
+    pub async fn bitcoin(&self) -> &BitcoinContainer {
+        self.bitcoin
+            .get_or_init(|| BitcoinContainer::from_containers(self))
+            .await
+    }
+
+    /// Get the database container
+    pub async fn database(&self) -> &DatabaseContainer {
+        self.database
+            .get_or_init(|| DatabaseContainer::from_containers(self))
+            .await
+    }
+}
+
+pub struct BitcoinContainer {
+    url: Url,
+    rpc_client: bitcoincore_rpc::Client,
+}
+
+impl BitcoinContainer {
+    pub async fn from_containers(containers: &TestContainers) -> Self {
+        let auth = bitcoincore_rpc::Auth::UserPass(
+            BITCOIN_CORE_RPC_USERNAME.to_string(),
+            BITCOIN_CORE_RPC_PASSWORD.to_string(),
+        );
+        let url = containers
+            .get_bitcoin_url()
+            .await
+            .expect("cannot get bitcoin url");
+
+        let rpc_client = bitcoincore_rpc::Client::new(url.as_ref(), auth)
+            .expect("cannot create bitcoin rpc client");
+
+        get_or_create_wallet(&rpc_client, BITCOIN_CORE_WALLET_NAME);
+        let faucet = Faucet::new(FAUCET_SECRET_KEY, AddressType::P2wpkh, &rpc_client);
+        faucet.track_address(FAUCET_LABEL);
+
+        let amount = rpc_client
+            .get_received_by_address(&faucet.address, None)
+            .unwrap();
+        if amount < Amount::from_int_btc(1) {
+            faucet.generate_blocks(MIN_BLOCKCHAIN_HEIGHT);
+        }
+
+        BitcoinContainer { url, rpc_client }
+    }
+
+    pub fn url(&self) -> &Url {
+        &self.url
+    }
+
+    pub fn rpc(&self) -> &bitcoincore_rpc::Client {
+        &self.rpc_client
+    }
+
+    pub fn get_faucet(&self) -> Faucet<'_> {
+        Faucet::new(FAUCET_SECRET_KEY, AddressType::P2wpkh, &self.rpc_client)
+    }
+
+    pub fn get_client(&self) -> BitcoinCoreClient {
+        self.url().try_into().expect("cannot create bitcoin client")
+    }
+}
+
+pub struct DatabaseContainer {
```

Comment:
Btw in my implementation, I opted to not containerize postgres per-test because our current solution with db-per-test-spawning is thread-safe and more efficient than spawning containers. I of course see the benefits of going all-out containers-per-test for consistency, but if the goal is speed then db-per-test will always win (with the downside of still needing an `integration-env-up`-like thingy.